File: Trunk/frameworks/runtime-src/Classes_backup/library/ByteArray/ByteArray.cpp

 #include "ByteArray.h"
#include "SpriteDefine.h"
int ByteArray::globalCPUEndian = -1;

ByteArray::ByteArray():
bytes(nullptr),
pos(0),
length(0),
endian(0)
{

};

ByteArray* ByteArray::create(int len)
{
    ByteArray* ba = new ByteArray();
    if (ba)
    {
        ba->autorelease();
        ba->pos = 0;
        ba->length = len;
    }
    else
    {
        delete ba;
        ba = NULL;
        return NULL;
    }

    //auto size = sizeof(ba->bytes);
    //ba->pos = 0;
    //ba->length = len;
    //ba->endian = 1;//endian;

    //if (globalCPUEndian == -1){
    //	globalCPUEndian  =  ByteArray::checkCPUEndian();
    //}

    return ba;
}
//
ByteArray* ByteArray::createWithNotify(CNotify* notify)
{
    ByteArray* ba = new ByteArray();
    if (ba)
    {
        ba->autorelease();
        ba->bytes = (BYTE*)notify->getData();
        ba->length = notify->getSize();
        return ba;
    }
    else
    {
        delete ba;
        ba = NULL;
        return NULL;
    }
    return ba;
}

ByteArray* ByteArray::createWithPacket(TCP_BufferRef* packet)
{
    ByteArray* ba = new ByteArray();
    if (ba)
    {
        ba->autorelease();
        ba->bytes = packet->cmd.cbBuffer;
        return ba;
    }
    else
    {
        delete ba;
        ba = NULL;
        return NULL;
    }
    return ba;
}

void ByteArray::destroy(ByteArray* ba) {
    delete ba;
}

//0: big 1:little
int ByteArray::checkCPUEndian() {
    union w
    {
        int a;
        char b;
    } c;
    c.a = 1;
    return  (c.b ==1 ? endianLittle : endianBig);
}

int ByteArray::getCPUEndian() {
    return globalCPUEndian;
}

bool ByteArray::readBool()
{
    bool b;
    char* p = (char*)&b;
    memcpy(p,this->bytes+this->pos,sizeof(b));
    this->pos += sizeof(b);
    return b;
}

void ByteArray::writeBool(bool value)
{
    char* p = (char*)&value;
    memcpy(this->bytes+this->pos,p,sizeof(value));
    this->pos += sizeof(value);
}

short ByteArray::readShort()
{
    short n;
    char* p = (char*)&n;
    memcpy(p,this->bytes+this->pos,sizeof(n));
    this->pos += 2;
    return n;
}

void ByteArray::writeShort(short value)
{
    char* p = (char*)&value;
    memcpy(this->bytes+this->pos,p,sizeof(value));
    this->pos += 2;
}

int ByteArray::readInt() {
    int n;
    char* p = (char*)&n;
    memcpy(p,this->bytes+this->pos,sizeof(int));
    this->pos += 4;
    return n;
}

void ByteArray::writeInt(int value)
{
    char* p = (char*)&value;
    memcpy(this->bytes+this->pos,p,sizeof(value));
    this->pos += 4;
}

float ByteArray::readFloat() {
    float n;
    char* p = (char*)&n;
    memcpy(p,this->bytes+this->pos,sizeof(float));
    this->pos += 4;
    return n;
}

void ByteArray::writeFloat(float value)
{
    char* p = (char*)&value;
    memcpy(this->bytes+this->pos,p,sizeof(value));
    this->pos += 4;
}

unsigned int ByteArray::readUINT() {
    unsigned int n;
    char* p = (char*)&n;
    memcpy(p,this->bytes+this->pos,sizeof(n));

    this->pos += 4;
    return n;
}

void ByteArray::writeUINT(unsigned int value)
{
    char* p = (char*)&value;
    memcpy(this->bytes+this->pos,p,sizeof(value));
    this->pos += 4;
}

unsigned short ByteArray::readWORD() {
    unsigned short n;
    char* p = (char*)&n;
    memcpy(p,this->bytes+this->pos,sizeof(n));
    this->pos += 2;
    return n;
}

void ByteArray::writeWORD(unsigned short value)
{
    char* p = (char*)&value;
    memcpy(this->bytes+this->pos,p,sizeof(value));
    this->pos += 2;
}

char ByteArray::readByte() {
    signed char val = this->bytes[this->pos];
    if (val > 127){
        val = val - 255;
    }
    this->pos += 1;
    return val;
}

void ByteArray::writeByte(char value)
{
    char* p = (char*)&value;
    memcpy(this->bytes+this->pos,p,sizeof(value));
    this->pos += 1;
}

unsigned char ByteArray::readBYTE() {
    unsigned char val = this->bytes[this->pos];
    this->pos += 1;
    return val;
}

void ByteArray::writeBYTE(unsigned char value)
{
    char* p = (char*)&value;
    memcpy(this->bytes+this->pos,p,sizeof(value));
    this->pos += 1;
}

unsigned long ByteArray::readDWORD()
{
    int n;
    char* p = (char*)&n;
    memcpy(p,this->bytes+this->pos,sizeof(n));
    this->pos += sizeof(n);
    return n;
}

void ByteArray::writeDWORD(int value)
{
    char* p = (char*)&value;
    memcpy(this->bytes+this->pos,p,sizeof(value));
    this->pos += sizeof(value);
}

LONGLONG ByteArray::readLONGLONG()
{
    long long n;
    char* p = (char*)&n;
    memcpy(p,this->bytes+this->pos,sizeof(n));
    this->pos += sizeof(n);
    return n;
}

void ByteArray::writeLONGLONG(long long value)
{
    char* p = (char*)&value;
    memcpy(this->bytes+this->pos,p,sizeof(value));
    this->pos += sizeof(value);
}

string ByteArray::readString(int len) {

    char* value = new char[len + 1];

    value[len] = 0;
    memcpy(value, this->bytes + this->pos, len);
    this->pos += len;

    string str(value);
    delete [] value;
    value = nullptr;

    return str;
}

string ByteArray::readTchar(int len)
{
    TCHAR* tchar = new TCHAR[len];
    memset(tchar,0,len);
    memcpy(tchar, this->bytes + this->pos, len*sizeof(TCHAR));
    this->pos += len*sizeof(TCHAR);

    string str(YYConverUnicodeToUtf8(tchar));
    delete[] tchar;
    tchar = nullptr;

    return str;
}

void ByteArray::writeTchar(const char* value,int len)
{
    TCHAR* tchar = new TCHAR[len];
    memset(tchar,0,len);
    YYConverUtf8ToUnicodeWithArray(tchar,len,value);
    memcpy(this->bytes+this->pos,tchar,len);
    this->pos += len*sizeof(TCHAR);

    delete[] tchar;
    tchar = nullptr;
}

string ByteArray::readTCHAR(int len)
{
	return this->readTchar(len);
};

void ByteArray::writeTCHAR(const char* value,int len)
{
	this->writeTchar(value,len);
}

int ByteArray::getPosition() {

    return this->pos;
}

void ByteArray::setPosition(int pos) {

    this->pos = pos;
}


int ByteArray::getLength() {
    return this->length;
}


int ByteArray::getBytesAvailable() {
    return this->length - this->pos;
}

File: Trunk/frameworks/runtime-src/Classes/lua/api/ByteArray.lua

--------------------------------
-- @module ByteArray
-- @extend Ref
-- @parent_module jt

--------------------------------
-- 
-- @function [parent=#ByteArray] writeShort 
-- @param self
-- @param #short value
-- @return ByteArray#ByteArray self (return value: ByteArray)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] readBool 
-- @param self
-- @return bool#bool ret (return value: bool)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] writeDouble 
-- @param self
-- @param #double value
-- @return ByteArray#ByteArray self (return value: ByteArray)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] readByte 
-- @param self
-- @return char#char ret (return value: char)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] writeLONGLONG 
-- @param self
-- @param #long long value
-- @return ByteArray#ByteArray self (return value: ByteArray)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] getLength 
-- @param self
-- @return int#int ret (return value: int)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] readFloat 
-- @param self
-- @return float#float ret (return value: float)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] readTchar 
-- @param self
-- @param #int len
-- @return string#string ret (return value: string)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] writeBool 
-- @param self
-- @param #bool value
-- @return ByteArray#ByteArray self (return value: ByteArray)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] readInt 
-- @param self
-- @return int#int ret (return value: int)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] readBYTE 
-- @param self
-- @return unsigned char#unsigned char ret (return value: unsigned char)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] writeFloat 
-- @param self
-- @param #float value
-- @return ByteArray#ByteArray self (return value: ByteArray)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] writeTchar 
-- @param self
-- @param #char value
-- @param #int len
-- @return ByteArray#ByteArray self (return value: ByteArray)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] writeUINT 
-- @param self
-- @param #unsigned int value
-- @return ByteArray#ByteArray self (return value: ByteArray)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] readTCHAR 
-- @param self
-- @param #int len
-- @return string#string ret (return value: string)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] writeDWORD 
-- @param self
-- @param #int value
-- @return ByteArray#ByteArray self (return value: ByteArray)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] readShort 
-- @param self
-- @return short#short ret (return value: short)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] writeWORD 
-- @param self
-- @param #unsigned short value
-- @return ByteArray#ByteArray self (return value: ByteArray)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] readDWORD 
-- @param self
-- @return unsigned long#unsigned long ret (return value: unsigned long)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] writeByte 
-- @param self
-- @param #char value
-- @return ByteArray#ByteArray self (return value: ByteArray)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] readLONGLONG 
-- @param self
-- @return long long#long long ret (return value: long long)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] readDouble 
-- @param self
-- @return double#double ret (return value: double)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] setPosition 
-- @param self
-- @param #int pos
-- @return ByteArray#ByteArray self (return value: ByteArray)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] getCPUEndian 
-- @param self
-- @return int#int ret (return value: int)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] readUINT 
-- @param self
-- @return unsigned int#unsigned int ret (return value: unsigned int)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] readString 
-- @param self
-- @param #int len
-- @return string#string ret (return value: string)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] getPosition 
-- @param self
-- @return int#int ret (return value: int)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] getBytesAvailable 
-- @param self
-- @return int#int ret (return value: int)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] writeBYTE 
-- @param self
-- @param #unsigned char value
-- @return ByteArray#ByteArray self (return value: ByteArray)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] writeInt 
-- @param self
-- @param #int value
-- @return ByteArray#ByteArray self (return value: ByteArray)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] writeTCHAR 
-- @param self
-- @param #char value
-- @param #int len
-- @return ByteArray#ByteArray self (return value: ByteArray)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] readWORD 
-- @param self
-- @return unsigned short#unsigned short ret (return value: unsigned short)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] destroy 
-- @param self
-- @param #ByteArray ba
-- @return ByteArray#ByteArray self (return value: ByteArray)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] create 
-- @param self
-- @param #int len
-- @return ByteArray#ByteArray ret (return value: ByteArray)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] createWithPacket 
-- @param self
-- @param #TCP_BufferRef packet
-- @return ByteArray#ByteArray ret (return value: ByteArray)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] checkCPUEndian 
-- @param self
-- @return int#int ret (return value: int)
        
--------------------------------
-- 
-- @function [parent=#ByteArray] createWithNotify 
-- @param self
-- @param #CNotify notify
-- @return ByteArray#ByteArray ret (return value: ByteArray)
        
return nil
